### ISSUE TITLE
"split features" of PostGIS layers become slow since 3.6

### ISSUE THREAD

**Issue by oliviersp** [qgis/QGIS#34326]
Since version 3.6 of Qgis, the digitalization tools “split Features” , is very slow to perform the cutting (several seconds). The more polygons the cutting line crosses, the longer the cutting execution time takes.
Configuration:
Qgis 3.6 or 3.8 or 3.10
PostgreSQL 10.6 database /POSTGIS="2.5.2 r17328 "
OS Windows 10 or Opensuse 15.1
This bug did not exist on Qgis 3.4 (Windows or Linux) for which cutting is very fast.

**Comment by Doctor-Who** [comment]
As it's a critical issue from our point of view. We can found thix fix.

**Comment by elpaso** [comment]
@Doctor-Who can you please share a project and a small DB SQL dump to reproduce the issue?

For commercial support you can contact any of the firms listed in  https://www.qgis.org/fr/site/forusers/commercial_support.html#qgis-commercial-support

**Comment by oliviersp** [comment]
Here the link to dowload a data base SQL to reproduce the issue.
https://framadrop.org/r/Mx1Sv6vHkc#L7qpVbXkzu/g2b249J12B4hVPzyX0eMTvFVyvESpxUY=
Sorry for the size of data, if i reduce it, there isn't slow-down.

**Comment by elpaso** [comment]
@oliviersp I'm testing your dump right now. I had to run "Fix geometries" algorithm before testing but I don't see any particular slowness, splitting the whole dataset from top to bottom takes roughly 20 seconds, but that is kind of expected given the large number of feature.

In other words, it doesn't look particularly slow to me, but I have a fast machine.

**Comment by oliviersp** [comment]
Could you try with this data please ?
https://framadrop.org/r/VsJam3oheH#J60L6DqTSdUP0QowHcThgmQvImRhGf/rxWZ+gpkwLgo=

**Comment by oliviersp** [comment]
This is the cutting line that i tried on the data :
![line_of_split_features](https://user-images.githubusercontent.com/29973752/74033516-a1769480-49b6-11ea-9acc-9935be5e5227.png)

**Comment by elpaso** [comment]
@oliviersp how am I supposed to find that particular feature?

**Comment by oliviersp** [comment]
gid = 264038
location : 1046606.28 , 6842771.43
But you can try with any polygon.

**Comment by oliviersp** [comment]
I export data in shapefile format. The same problem occurs on Qgis 3.6 or 3.8 or 3.10.

**Comment by oliviersp** [comment]
sorry, gid = 60716 
and location  : 547454.9,6505538.6
![line_of_split_features](https://user-images.githubusercontent.com/29973752/74035879-70e52980-49bb-11ea-834a-75191a423dea.png)

**Comment by elpaso** [comment]
It takes ~2 secs here to split in 7 parts. 
With a shapefile export it's less than sec.

**Comment by oliviersp** [comment]
How do you explain that it takes 7 seconds on qgis3.6 or 3.8 or 3.10 and it takes less than 1 sedond on qgis3.4 (on the same PC) ?

**Comment by elpaso** [comment]
@olivierdalang I can't explain. 

But the logic after 3.4 has been improved to handle a lot of corner cases like backend-side defaults, constraints, unique values and so on. The fact is that they should not kick in in case of a shapefile (that has no backend-side constraints).

In any case, I see less than 1 sec on current master debug build with 10 cuts on shapefiles, an ~2 seconds with postgis.

I just tested 3.4 and I've got the same results, actually probably a bit faster because it's not a debug build.

**Comment by Doctor-Who** [comment]
@elpaso As several people are working on same layer and some triggers are working in real time, we cannot on filesystem but our database server is very efficient. 
We will test on latest master thanks for those explanations.

**Comment by gioman** [comment]
> How do you explain that it takes 7 seconds on qgis3.6 or 3.8 or 3.10 and it takes less than 1 sedond on qgis3.4 (on the same PC) ?

@oliviersp just tested on 3.10.2 on Linux and it took not much more than 1 sec to split that geometry with that split line.

**Comment by SrNetoChan** [comment]
@oliviersp can you try to use a clean profile before you do that operation? It can be something related to your settings or a plugin that it's affecting all versions except 3.4. maybe 3.4 is not using the default profile and the others are.

**Comment by oliviersp** [comment]
I tried on Qgis 3.10.2 on Linux with a new profile. There is no improvement.

**Comment by gioman** [comment]
> I tried on Qgis 3.10.2 on Linux with a new profile. There is no improvement.

@oliviersp does it happen also if you load that layer alone in completely clean project?

**Comment by oliviersp** [comment]
> 
> 
> > I tried on Qgis 3.10.2 on Linux with a new profile. There is no improvement.
> 
> @oliviersp does it happen also if you load that layer alone in completely clean project?

Yes, the layer is alone in a clean project.

**Comment by gioman** [comment]
> Yes, the layer is alone in a clean project.

@oliviersp remote server? local network?

**Comment by oliviersp** [comment]
The database is on a local network server.

**Comment by gioman** [comment]
> The database is on a local network server.

@oliviersp could you try if it makes any difference importing the table dump on a different endpoint, for example one installed locally on your machine?

**Comment by oliviersp** [comment]
In a local Postgres database on SSD, it's a little bit faster (between 5 and 6 seconds against 10 seconds).
In the very case the layer contains a great number of polygons and the cutting line crosses several of these polygons, then, we note, the cutting execution time will be very long.

**Comment by oliviersp** [comment]
Could you try with this data please ?
https://framadrop.org/r/DOWSJhoHEx#lApERwY00Si5FfM2rgVqHT0VnLPdwwgxCSXQkWcTiTw=

**Comment by oliviersp** [comment]
![line_of_split_features](https://user-images.githubusercontent.com/29973752/74161593-7773db00-4c1f-11ea-8e0b-2f434ee2c6ac.png)

at location : 771030.44,4860111.74

**Comment by gioman** [comment]
> Could you try with this data please ?

@oliviersp I cannot cleanly restore this dump: ERROR:  relation "prod_d67_extrait2" does not exist

**Comment by oliviersp** [comment]
I create a new extract of data. 
https://framadrop.org/r/N_Kaarr3Ba#371y3fIygmNxTptviAlesnEK018QfeNq2E3VhaCwpxQ=
location : 246698.9,5242970.5
It takes 10 seconds to cut.
![line_of_split_features_extract_2](https://user-images.githubusercontent.com/29973752/74314784-5d004580-4d76-11ea-816e-c9dfa72650f1.png)

**Comment by elpaso** [comment]
Just tested again latest master with your latest dataset and the split operation (16 cuts) is almost immediate (less than 1 second).

**Comment by oliviersp** [comment]
What are your versions of Postgres and Postgis ?

**Comment by elpaso** [comment]
psql (12.1 (Ubuntu 12.1-1.pgdg18.04+1), server 10.11 (Ubuntu 10.11-1.pgdg18.04+1))
```
qgis_tests=# select postgis_version();
            postgis_version            
---------------------------------------
 2.4 USE_GEOS=1 USE_PROJ=1 USE_STATS=1
```

**Comment by gioman** [comment]
@oliviersp on a system equal to Alessandro's it takes 2/3 seconds on my laptop (QGIS 3.10.2), that is 8 years old.

**Comment by oliviersp** [comment]
gioman, what are your versions of Postgres and Postgis ?

**Comment by gioman** [comment]
same as Alessandro's (I believe we are both on Ubuntu 18.04).

**Comment by Doctor-Who** [comment]
Ok we have to upgrade to PostGreSQL 12. Thanks to all for your help.

**Comment by Doctor-Who** [comment]
We've test on Docker Database optimize with PgTune with PostGreSQL 12 same problem.
@gioman @elpaso From my point of view, if the network will be guilty, then in 3.4 it won't be so efficient ? All version will slow down and not only QGIS 3.6 and later ? 
The database server got 4x10Gbits network and load is around 20-25% as it has 56 threads and 128 GB RAM with 1.2 TiB SSD storage

**Comment by gioman** [comment]
@oliviersp @elpaso I uploaded the last dataset to a remote server (not even in the same local network where my laptop is) and yes, I realized that splitting (and maybe also other operations) is much faster on 3.4.15 than 3.10.2. I'll leave the call to reopen to a developer, maybe this is because of some underlying change by design.

**Comment by roya0045** [comment]
Could this be a side effect of some of the fixes done to PostGIS?

**Comment by saberraz** [comment]
> Here the link to download a data base SQL to reproduce the issue.
> https://framadrop.org/r/Mx1Sv6vHkc#L7qpVbXkzu/g2b249J12B4hVPzyX0eMTvFVyvESpxUY=
> Sorry for the size of data, if i reduce it, there isn't slow-down.

Would it be possible to re-upload the data? The link has expired.

**Comment by oliviersp** [comment]
Here another link to download a data base SQL to reproduce the issue.
https://framadrop.org/r/a7J3ZSWsiE#UAXcNeY6261Yl9kTRDNKseZ+sb2wPnWaWWS1UZgJkDM=

**Comment by oliviersp** [comment]
I tried on Qgis 3.12.0 on OpensSUSE with a new profile. There is no improvement.

**Comment by saberraz** [comment]
@olivierdalang would it be possible to have it as compressed binary format? I had a quick look and could not restore the backup.

**Comment by olivierdalang** [comment]
@saberraz I guess you wanted to ping @oliviersp ?

**Comment by saberraz** [comment]
hehehe...indeed...sorry for waking you up! @oliviersp

**Comment by oliviersp** [comment]
You can try this one (TAR)
 https://framadrop.org/r/FjWD1akuYd#OKfEvNdbGKLhHWz/SXESs+75UxhcaZI/fuSNNyG2R+8=

**Comment by gioman** [comment]
> You can try this one (TAR)

@oliviersp best is always pg_dump custom (-Fc) binary format

**Comment by saberraz** [comment]
OK ...managed to restore the database and all worked fine on mine too:
QGIS master
psql (PostgreSQL) 11.7 
Ubuntu 19.10
![Peek 2020-03-12 15-22](https://user-images.githubusercontent.com/2649047/76537385-97dcc280-6475-11ea-8282-687b3bd02475.gif)

**Comment by gioman** [comment]
> OK ...managed to restore the database and all worked fine on mine too:

@saberraz if you don't load the data on a remote server you are unlikely to see any difference in performance of this operation, see: https://github.com/qgis/QGIS/issues/34326#issuecomment-585328814

**Comment by saberraz** [comment]
In that case, it is difficult to replicate if we don't have access to the production server. As others mentioned, it could be due to a number of things.

**Comment by gioman** [comment]
> In that case, it is difficult to replicate if we don't have access to the production server. As others mentioned, it could be due to a number of things.

@saberraz yes, but I did replicate it on a production server.

**Comment by oliviersp** [comment]
I send another data in a few minutes.

**Comment by oliviersp** [comment]
You can try with this data 
https://framadrop.org/r/ckjbNubz0t#BRDM8rLZ1jftosWy0uZFFa+ZWfy3UtJLweasqI8fLvU=

**Comment by gioman** [comment]
This is still an issue, see https://github.com/qgis/QGIS/pull/37295#issuecomment-709014889

**Comment by Doctor-Who** [comment]
Yes, it's still a issue even in latest QGIS version, I'm funding Lutra for that but it's an hard work to do.

**Comment by wonder-sk** [comment]
I am going to close this ticket because there are at least two issues mixed up here. One issue was with split of a complex polygon (already fixed in #37295) and another one when layer contains lots of features - I have raised a new ticket for it: #40860.
